automation/JSInstrumentation.py

```
@@ -0,0 +1,204 @@
+    If the item is a string:
+      - Is it a shortcut e.g. "fingerprinting"
+      - Is it an object (verified by mdn-browser-compat)
+      - If neither, reject
+    If the item is an object:
+      - Must only have one property
+      - The value may be a list or a LogSettings object
+      - If the value is a list, it is transformed into the
+        propertiesToInstrument property of a new LogSettings object.
+    We must also create the instrumentedName value.
+    """
+    obj = None
+    instrumentedName = None
+    logSettings = get_default_log_settings()
+    if isinstance(request, str):
+        obj, instrumentedName = _handle_obj_string(request)
+    elif isinstance(request, dict):
+        assert len(request.keys()) == 1
+        req = list(request.keys())[0]
+        props = request[req]
+        obj, instrumentedName = _handle_obj_string(req)
+        if isinstance(props, list):
+            logSettings['propertiesToInstrument'] = props
+        elif isinstance(props, dict):
+            for k, v in props.items():
+                logSettings[k] = v
+        else:
+            raise RuntimeError('Invalid settings for object')
+    else:
+        raise RuntimeError('Invalid input')
+    return {
+        'object': obj,
+        'instrumentedName': instrumentedName,
+        'logSettings': logSettings
+    }
+
+
+def convert_browser_params_to_js_string(js_instrument_modules):
+    """
+    We accept a list. From the list we need to parse each item.
+    Examples of the requests we accept.
+    // Shortcut
+    "fingerprinting",
+    // APIs
+    {"XMLHttpRequest": {"badSetting": 1}},
+    {"XMLHttpRequest": {"excludedProperties": ["send"]}},
+    {"Prop1": ["hi"], "Prop2": ["hi2"]},
+    {"XMLHttpRequest": ["send"]},
+    "Storage",
+    // Specific instances on window
+    {"window.document": ["cookie", "referrer"]},
+    {"window": ["name", "localStorage", "sessionStorage"]}
+    """
+    assert isinstance(js_instrument_modules, list)
```

You shouldn't use assert here. This method is potentially parsing user input and we'll always want to throw an error if there's an issue with the parameters (i.e.,`ValueError`). Assert can be [optimized out](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement); it should only be used for debugging.

---

It looks like this applies to all of the assert statements here; if a user is constructing their own settings they may cause any of these mistakes accidentally.

---

Resolved in 520c782